Add support for response writing to RequestDelegateGenerator

captainsafia opened the pull request:
Addresses https://github.com/dotnet/aspnetcore/issues/46276 and resolves #45524.

- Share WellKnownTypes infrastructure between analyzers and generator
- Emit diagnostics from RequestDelegateGenerator for unsupported syntax
- Add support for writing more response types to generator
- Dedupe strongly-typed endpoints generated from the same delegate type
- Use RequestDelegateGenerator in MinimalSample
- Fix nullability warnings in generated source when referencing `RequestDelegateFactoryOptions.EndpointBuilder`

captainsafia committed:
Add support for response writing to RequestDelegateGenerator

JamesNK commented on src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs:
```diff
@@ -0,0 +1,203 @@
+
+internal static class WellKnownTypeData
+{
+    public enum WellKnownType
```
Does this need to be nested?

captainsafia replied:
Not necessarily. It was largely a stylistic choice to put the associated data in the same class. To minimize the code delta, I ended up using an alias in the existing analyzers/codefixers but it felt nicer to capture these types in class instead of having them on the top-level namespace.

halter73 replied:
Wouldn't leaving this unnested minimize the code delta by removing the `using WellKnownType = WellKnownTypeData.WellKnownType;`?

JamesNK commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/Multiple_MapAction_NoParam_StringReturn.generated.txt:
```diff
@@ -0,0 +1,225 @@
+
+    file static class GeneratedRouteBuilderExtensionsCore
+    {
+
```
super nit: Is there a way to not start the type with a blank line? My OCD is triggered 😄

captainsafia replied:
Hm, I think we might be able to remove the newline in the interpolated string `{GetGenericThunks()}{GetThunks()}` and change the strings emitted from those methods to get around it.

In the long-term, we might be able to take advantage of the `SyntaxTree.NormalizeWhitespace` feature to correct the indentation for the outputted source. I wasn't able to use it successfully here because it has a really hard time formatting the Dictionary of source key to tuples of funcs pairing. As we evolve the code, we might be able to get rid of this (e.g. once the interceptors feature lands) and we can do all the formatting at once instead of doing the bookkeeping for the indents here.

A related issue on the Roslyn side is: https://github.com/dotnet/roslyn/issues/43821

JamesNK commented on src/Http/Http.Extensions/gen/Microsoft.AspNetCore.Http.Generators.csproj:
```diff
@@ -20,9 +20,13 @@
     <Compile Include="$(SharedSourceRoot)IsExternalInit.cs" LinkBase="Shared" />
+    <Compile Include="$(SharedSourceRoot)RoslynUtils\BoundedCacheWithFactory.cs" LinkBase="Shared" />
+    <Compile Include="$(SharedSourceRoot)RoslynUtils\CodeWriter.cs" LinkBase="Shared" />
+    <Compile Include="$(SharedSourceRoot)RoslynUtils\WellKnownTypes.cs" LinkBase="Shared" />
```
I think sharing code is ok for now, but doing this results in two caches. I think we should investigate whether we can have a shared library project output to the analyzers directory that generators, analyzers, and code fixers can reference.

For route pattern parsing, there is a lot more code, and we don't want to duplicate parsing and caching route patterns twice (once in analyzer assembly, once in generators assembly)

captainsafia replied:
> I think sharing code is ok for now, but doing this results in two caches. I think we should investigate whether we can have a shared library project output to the analyzers directory that generators, analyzers, and code fixers can reference.

This is an interesting proposal. @chsienki do you happen to know if there is a pattern for sharing a symbol cache across analyzers/generators in the same compilation process?

> For route pattern parsing, there is a lot more code, and we don't want to duplicate parsing and caching route patterns twice (once in analyzer assembly, once in generators assembly)

Roger that! I captured that in https://github.com/dotnet/aspnetcore/issues/46342 so we can start discussing what that might look like.

eerhardt commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs:
```diff
@@ -49,15 +60,65 @@ public static string EmitVerb(Endpoint endpoint)
+        else
+        {
+            code.WriteLine($"""httpContext.Response.ContentType ??= "{endpoint.Response.ContentType}";""");
+            if (endpoint.Response.IsAwaitable)
+            {
+                code.WriteLine("var result = await handler();");
+                code.WriteLine(endpoint.EmitResponseWritingCall());
+            }
+            else
+            {
+                code.WriteLine("var result = handler();");
+                code.WriteLine("return GeneratedRouteBuilderExtensionsCore.ExecuteObjectResult(result, httpContext);");
+            }
```
`EmitResponseWritingCall` handles `IsAwaitable()`. So I think this can be refactored to just:

```suggestion
            code.WriteLine(endpoint.Response.IsAwaitable
                ? "var result = await handler();"
                : "var result = handler();");
            code.WriteLine(endpoint.EmitResponseWritingCall());
```

eerhardt commented on src/Http/Http.Extensions/gen/GeneratorSteps.cs:
```diff
@@ -0,0 +1,9 @@
+
+internal class GeneratorSteps
+{
+    internal const string EndpointsStep = "EndpointModel";
```
Is it weird that the name of the variable and the string doesn't match?

eerhardt commented on src/Http/Http.Extensions/gen/GeneratorSteps.cs:
```diff
@@ -0,0 +1,9 @@
+
+internal class GeneratorSteps
+{
+    internal const string EndpointsStep = "EndpointModel";
+    internal const string EndpointsWithoutDiagnosicsStep = "EndpointsWithoutDiagnostics";
```
```suggestion
    internal const string EndpointsModel = nameof(EndpointModel);
    internal const string EndpointsWithoutDiagnosicsStep = nameof(EndpointsWithoutDiagnostics);
```

eerhardt commented on src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:
```diff
@@ -39,83 +41,141 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
+                {
+                    unchecked
+                    {
+                        var hashCode = SymbolEqualityComparer.Default.GetHashCode(endpoint.Response.ResponseType);
+                        hashCode = (hashCode * 397) ^ endpoint.Response.IsAwaitable.GetHashCode();
+                        hashCode = (hashCode * 397) ^ endpoint.Response.IsVoid.GetHashCode();
+                        hashCode = (hashCode * 397) ^ endpoint.HttpMethod.GetHashCode();
+                        return hashCode;
```
We could use the [Microsoft.Bcl.HashCode](https://nuget.org/packages/Microsoft.Bcl.HashCode) package here.

```suggestion
                        return HashCode.Combine(
                            SymbolEqualityComparer.Default.GetHashCode(endpoint.Response.ResponseType),
                            endpoint.Response.IsAwaitable,
                            endpoint.Response.IsVoid,
                            endpoint.HttpMethod);
```

halter73 commented on src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:
```diff
@@ -39,83 +41,141 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
+            {
+                var dedupedByDelegate = endpoints.Distinct(new LambdaComparer<Endpoint>((a, b) =>
+                {
+                    if (a.Response.IsAwaitable == b.Response.IsAwaitable &&
+                        a.Response.IsVoid == b.Response.IsVoid &&
+                        SymbolEqualityComparer.Default.Equals(a.Response.ResponseType, b.Response.ResponseType) &&
```
```suggestion
                    if (a.Response.WrappedResponseType == b.Response.WrappedResponseType && 
```

The call `EmitHandlerDelegateType()` uses the `WrappedResponseType`, so we need to compare those. Currently, the test case below only produces a MapGet overload for `Func<Task<string>>` and not `Func<ValueTask<string>>`.

```csharp
    [Fact]
    public async Task Multiple_MapAction_NoParam_TaskAndValueTaskReturn()
    {
        var source = """
app.MapGet("/en", () => Task.FromResult("Hello world!"));
app.MapGet("/es", () => new ValueTask<string>("Hola mundo!"));
""";
        var (_, compilation) = RunGenerator(source);

        await VerifyAgainstBaselineUsingFile(compilation);
    }
```

captainsafia replied:
Good scenario! I added it to the tests.

halter73 commented on src/Framework/AspNetCoreAnalyzers/src/Analyzers/Http/HeaderDictionaryAddAnalyzer.cs:
```diff
@@ -9,6 +9,8 @@
 
 namespace Microsoft.AspNetCore.Analyzers.Http;
 
+using WellKnownType = WellKnownTypeData.WellKnownType;
```
Nit: Can we put this with the other usings?

halter73 commented on src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypeData.cs:
```diff
@@ -0,0 +1,203 @@
+
+internal static class WellKnownTypeData
+{
+    public enum WellKnownType
```
Or can we just go back to not nesting this in another class so we can get rid of all the extra `using WellKnownType = WellKnownTypeData.WellKnownType;` lines?

eerhardt commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs:
```diff
@@ -0,0 +1,82 @@
+
+    public override int GetHashCode()
+    {
+        unchecked
+        {
+            var hashCode = HttpMethod.GetHashCode();
+            hashCode = (hashCode * 397) ^ Route.GetHashCode();
+            hashCode = (hashCode * 397) ^ Response.GetHashCode();
+            hashCode = (hashCode * 397) ^ Diagnostics.GetHashCode();
+            hashCode = (hashCode * 397) ^ Location.GetHashCode();
+            hashCode = (hashCode * 397) ^ Operation.GetHashCode();
+            return hashCode;
+        }
```
> but it's unavailable in netstandard2.0

https://www.nuget.org/packages/Microsoft.Bcl.HashCode makes it available.

captainsafia replied:
> https://www.nuget.org/packages/Microsoft.Bcl.HashCode makes it available.

Nice! It looks like we have this referenced in our dependencies already as well.

mitchdenny commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/StaticRouteHandlerModel.Emitter.cs:
```diff
@@ -49,15 +60,65 @@ public static string EmitVerb(Endpoint endpoint)
-                }
-""";
+        var code = new CodeWriter(new StringBuilder());
+        code.Indent(5);
```
When @captainsafia was getting @brunolins16 up to speed yesterday we were discussing the ```CodeWriter``` in general. My feedback was that using the CodeWriter rather than simple string concatenation/interpolation actually made it harder to reason over the generated code. I'd rather see us drop the CodeWriter (at the expense of indentation) and then either do some code formatting at the end before handing it back to Roslyn or just leave it a little messy. We can also do a Format Document on the generated code after build if we need to read it.

eerhardt replied:
The Regex source generator uses `IndentedTextWriter`. Seems to work really well there. You just `writer.Indent++` and `--` when you are done.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs#L30

captainsafia committed:
React to feedback and add incrementality tests

captainsafia commented:
🆙 📅 
- Removed `CodeWriter` in favor of string literals with indentation in code sources
- Added incrementality tests to validate behavior if Endpoint's comparable implementation
- Added and consume `Microsoft.Bcl.HashCode` API as a dependency
- Tweaked smaller pieces of feedback

captainsafia commented on src/Http/samples/MinimalSample/MinimalSample.csproj:
```diff
@@ -3,6 +3,8 @@
     <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
     <Nullable>enable</Nullable>
+    <EnableRequestDelegateGenerator>true</EnableRequestDelegateGenerator>
+    <EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>
```
I don't use Visual Studio so I'm not able to take advantage of any shortcuts there.

> Do we need this all the time?

I found it helpful to have on while debugging as it makes it easier to examine the generated code in another format. There's no con to having it so I figured it was helpful.

eerhardt replied:
> I don't use Visual Studio

![image](https://user-images.githubusercontent.com/8291187/216646395-ff41a201-db1c-4df2-a7d9-efb87a1ae9c1.png)

captainsafia commented on src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Endpoint.cs:
```diff
@@ -0,0 +1,82 @@
+            return endpoint.HttpMethod.Equals(HttpMethod, StringComparison.OrdinalIgnoreCase) ||
+                endpoint.Location.Item1.Equals(Location.Item1, StringComparison.OrdinalIgnoreCase) ||
+                endpoint.Location.Item2.Equals(Location.Item2) ||
+                endpoint.Response.Equals(Response);
```
Ooof! Good catch on the `||` vs `&&`. I added incrementality tests that'll catch things like this in the future.

captainsafia committed:
Route patterns don't factor into incrementality check

halter73 reviewed (approved):
I've got a draft PR based on this branch open at #46407. Now I just want to minimize my merge conflicts! 😆 Looks good though!

captainsafia committed:
Address feedback from review

captainsafia commented:
Merging to unblock forthcoming PRs. Happy to address any other feedback in a followup PR. :shipit:
